engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/EngineJobHashingActor.scala

```diff
@@ -1,53 +1,112 @@
+import wdl4s.values.WdlFile
+import HashValue.StringMd5er
 
 /**
   * Over time will emit up to two messages back to its parent:
   *  * (if read enabled): Either a CacheHit(id) or CacheMiss message
   *  * (if write enabled): A CallCacheHashes(hashes) message
   */
 case class EngineJobHashingActor(jobDescriptor: BackendJobDescriptor,
-                                 fileHasherActor: ActorRef,
+                                 backendSpecificHasherActor: ActorRef,
+                                 backendName: String,
                                  mode: CallCachingActivity) extends LoggingFSM[EJHAState, EJHAData] with ActorLogging {
 
   val receiver = context.parent
 
   initializeEJHA()
 
   when(DeterminingHitOrMiss) {
-    case Event(hashResultMessage: HashResultMessage, _) => lookupRelevantCacheResults(hashResultMessage.hashes)
-    case Event(newCacheResults: CacheResultMatchesForHashes, _) => checkWhetherHitOrMissIsKnownThenTransition(stateData.intersectCacheResultIds(newCacheResults))
+    case Event(hashResultMessage: SuccessfulHashResultMessage, _) =>
+      // This is in DeterminingHitOrMiss, so use the new hash results to search for cache results.
+      // Also update the state data with these new hash results.
+      val newHashResults = hashResultMessage.hashes
+      findCacheResults(newHashResults)
+      stay using updateStateDataWithNewHashResults(newHashResults)
+    case Event(newCacheResults: CacheResultMatchesForHashes, _) =>
+      checkWhetherHitOrMissIsKnownThenTransition(stateData.intersectCacheResults(newCacheResults))
+    case Event(PlaceholderHashKeyExpansion(placeholderHashKey, newHashKeys), _) => checkWhetherHitOrMissIsKnownThenTransition(stateData.replacePlaceholderHashKey(placeholderHashKey, newHashKeys))
   }
 
   when(GeneratingAllHashes) {
-    case Event(hashResultMessage: HashResultMessage, _) => checkWhetherAllHashesAreKnownAndTransition(stateData.withNewKnownHashes(hashResultMessage.hashes))
+    case Event(hashResultMessage: SuccessfulHashResultMessage, _) => checkWhetherAllHashesAreKnownAndTransition(stateData.withNewKnownHashes(hashResultMessage.hashes))
+    case Event(PlaceholderHashKeyExpansion(placeholderHashKey, newHashKeys), _) => checkWhetherAllHashesAreKnownAndTransition(stateData.replacePlaceholderHashKey(placeholderHashKey, newHashKeys))
     case Event(CacheResultMatchesForHashes(_, _), _) => stay // Don't care; we already know the hit/miss status. Ignore this message
   }
 
+  whenUnhandled {
+    case Event(failure: CacheResultLookupFailure, _) =>
+      // Crash and let the supervisor deal with it.
+      throw new RuntimeException("Failure looking up call cache results", failure.reason)
+    case Event(HashingFailedMessage(hashKey, reason), _) =>
+      receiver ! HashError(new Exception(s"Unable to generate ${hashKey.key} hash. Caused by ${reason.getMessage}", reason))
```

Comment:
We want to continue but treat it like a cache miss and not write anything new to the cache.

I could **probably** do the same with crashes and state-specific recovery options in the EJEA... but this just feels so much neater